File: src/ControlPanel.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;

namespace OpusScreen
{
    /// <summary>
    /// Fenetre de reglages.
    ///
    /// Sept pages plutot qu'un long formulaire : avec une quarantaine de reglages,
    /// tout montrer d'un coup rendrait l'application inutilisable pour le geste
    /// quotidien, qui se resume a bouger un curseur. La premiere page porte ce geste ;
    /// les six autres attendent qu'on aille les chercher.
    ///
    /// Cette classe n'assemble que la coquille : chaque page connait son domaine et
    /// rien d'autre.
    /// </summary>
    public class ControlPanel : Form
    {
        private readonly Settings _s;
        private readonly DisplayController _display;
        private readonly Action _push;

        private SideNav _nav;
        private Panel _content, _header, _footer;
        private Label _title, _subtitle, _status;
        private DarkButton _pauseButton;

        private readonly List<SettingsPage> _pages = new List<SettingsPage>();

        public PageWelcome WelcomePage;
        public PageDisplay Display1;
        public PageColor ColorPage;
        public PageColorBlind ColorBlindPage;
        public PageVision VisionPage;
        public PageAuto AutoPage;
        public PageApps AppsPage;
        public PageScreens ScreensPage;
        public PageComfort ComfortPage;
        public PageHotkeys HotkeysPage;
        public PageAdvanced AdvancedPage;

        /// <summary>
        /// Rang de l'onglet Daltonisme, ouvert depuis le menu de la zone de
        /// notification.
        ///
        /// Calcule depuis la liste, et non ecrit a la main : la premiere version
        /// etait une constante, et le jour ou un onglet s'est glisse avant celui-la,
        /// le menu a ouvert la page voisine sans que rien ne le signale.
        /// </summary>
        public int ColorBlindPageIndex { get { return _pages.IndexOf(ColorBlindPage); } }

        /// <summary>Rang de la page Vision (basse vision, pointeur, lecture).</summary>
        public int VisionPageIndex { get { return _pages.IndexOf(VisionPage); } }

        /// <summary>Largeur de la colonne de navigation, referencee par la mise en page de l'en-tete.</summary>
        private static int NavWidth { get { return Theme.Px(196); } }

        /// <summary>Declenche la suspension ou la reprise depuis le bouton du pied de page.</summary>
        public Action TogglePause;

        public ControlPanel(Settings s, DisplayController display, Action push)
        {
            _s = s;
            _display = display;
            _push = push;
            BuildUi();
            SyncAll();
        }

        // ------------------------------------------------------------------ construction

        private void BuildUi()
        {
            Text = "OpusScreen";
            BackColor = Theme.Bg;
            ForeColor = Theme.Fg;
            Font = Theme.Body;
            FormBorderStyle = FormBorderStyle.Sizable;
            MaximizeBox = false;
            StartPosition = FormStartPosition.Manual;

            // La fenetre se montre dans la barre des taches, sous l'icone de
            // l'application : c'est la que Windows regroupe ce qui est epingle.
            ShowInTaskbar = true;
            AppIcon.ApplyTo(this);
            ClientSize = new Size(Theme.Px(806), Theme.Px(660));
            MinimumSize = new Size(Theme.Px(760), Theme.Px(560));
            KeyPreview = true;

            // ---------------- en-tete ----------------
            _header = new Panel();
            _header.Dock = DockStyle.Top;
            _header.Height = Theme.Px(66);
            _header.BackColor = Theme.Bg;
            Controls.Add(_header);

            _title = new Label();
            _title.Font = Theme.Title;
            _title.ForeColor = Theme.Fg;
            _title.AutoSize = false;
            _header.Controls.Add(_title);

            _subtitle = new Label();
            _subtitle.Font = Theme.Small;
            _subtitle.ForeColor = Theme.Dim;
            _subtitle.AutoSize = false;
            _header.Controls.Add(_subtitle);

            _status = new Label();
            _status.Font = Theme.Small;
            _status.ForeColor = Theme.Dim;
            _status.AutoSize = false;
            _status.TextAlign = ContentAlignment.MiddleRight;
            _header.Controls.Add(_status);

            // ---------------- pied ----------------
            _footer = new Panel();
            _footer.Dock = DockStyle.Bottom;
            _footer.Height = Theme.Px(52);
            _footer.BackColor = Theme.Bg;
            Controls.Add(_footer);

            _pauseButton = new DarkButton();
            _pauseButton.Text = "Suspendre";
            _pauseButton.SetBounds(Theme.Px(20), Theme.Px(10), Theme.Px(116), Theme.MinTarget);
            _pauseButton.Click += delegate { if (TogglePause != null) TogglePause(); };
            _footer.Controls.Add(_pauseButton);

            DarkButton reset = new DarkButton();
            reset.Text = "Tout reinitialiser";
            reset.SetBounds(Theme.Px(144), Theme.Px(10), Theme.Px(132), Theme.MinTarget);
            reset.Click += OnReset;
            _footer.Controls.Add(reset);

            Label panic = new Label();
            panic.Text = "Secours : Ctrl + Alt + Maj + R retablit un ecran normal, meme si l'application ne repond plus.";
            panic.Font = Theme.Small;
            panic.ForeColor = Theme.Faint;
            panic.AutoSize = false;
            panic.TextAlign = ContentAlignment.MiddleRight;
            panic.Anchor = AnchorStyles.Top | AnchorStyles.Left | AnchorStyles.Right;
            panic.SetBounds(Theme.Px(286), Theme.Px(10), ClientSize.Width - Theme.Px(306), Theme.MinTarget);
            _footer.Controls.Add(panic);

            // ---------------- navigation ----------------
            // Le nom du produit vit dans la colonne de navigation, pas dans l'en-tete :
            // l'en-tete est ancre a droite du menu et son espace revient au titre de page.
            _nav = new SideNav();
            _nav.Dock = DockStyle.Left;

            // 196 plutot que 180 : le logo prend la place a gauche du nom, et
            // « OpusScreen » ressortait tronque en « OpusScree ».
            _nav.Width = NavWidth;
            Controls.Add(_nav);

            // Le logo tient a gauche du nom plutot qu'au-dessus : la colonne est
            // etroite, et chaque ligne prise en hauteur est une ligne de moins pour
            // les onglets.
            Image logo = AppIcon.Logo;
            int textLeft = Theme.Px(20);
            if (logo != null)
            {
                PictureBox mark = new PictureBox();
                mark.Image = logo;
                mark.SizeMode = PictureBoxSizeMode.Zoom;
                mark.BackColor = Color.Transparent;
                mark.SetBounds(Theme.Px(14), Theme.Px(15), Theme.Px(38), Theme.Px(32));
                _nav.Controls.Add(mark);
                textLeft = Theme.Px(58);
            }

            Label brand = new Label();
            brand.Text = "OpusScreen";
            brand.Font = Theme.Title;
            brand.ForeColor = Theme.Accent;
            brand.AutoSize = false;
            brand.BackColor = Color.Transparent;
            // Hauteur mesuree sur la police : ecrite en dur a 24 pixels, la marque
            // sortait deja tronquee en « OpusScree » sur un ecran a 150 %.
            int hMarque = TextRenderer.MeasureText("Ag", Theme.Title).Height + Theme.Px(2);
            brand.SetBounds(textLeft, Theme.Px(14), NavWidth - textLeft - Theme.Px(6), hMarque);
            _nav.Controls.Add(brand);

            Label version = new Label();
            // Lue dans l'assemblage plutot qu'ecrite ici : la version affichee avait
            // deja pris du retard sur la version reelle.
            Version v = System.Reflection.Assembly.GetExecutingAssembly().GetName().Version;
            version.Text = "version " + v.Major + "." + v.Minor;
            version.Font = Theme.Small;
            version.ForeColor = Theme.Faint;
            version.AutoSize = false;
            version.BackColor = Color.Transparent;
            int hVersion = TextRenderer.MeasureText("Ag", Theme.Small).Height + Theme.Px(2);
            version.SetBounds(textLeft + 1, Theme.Px(14) + hMarque, NavWidth - textLeft - Theme.Px(6), hVersion);

            // Les onglets commencent sous la marque, quelle que soit sa taille.
            _nav.TopOffset = Theme.Px(14) + hMarque + hVersion + Theme.SpaceSm;
            _nav.Controls.Add(version);

            // ---------------- contenu ----------------
            _content = new Panel();
            _content.Dock = DockStyle.Fill;
            _content.BackColor = Theme.Card;
            Controls.Add(_content);

            _content.BringToFront();

            // ---------------- pages ----------------
            WelcomePage = new PageWelcome(_s, _display, _push);
            Display1 = new PageDisplay(_s, _display, _push);
            ColorPage = new PageColor(_s, _display, _push);
            ColorBlindPage = new PageColorBlind(_s, _display, _push);
            VisionPage = new PageVision(_s, _display, _push);
            AutoPage = new PageAuto(_s, _display, _push);
            AppsPage = new PageApps(_s, _display, _push);
            ScreensPage = new PageScreens(_s, _display, _push);
            ComfortPage = new PageComfort(_s, _display, _push);
            HotkeysPage = new PageHotkeys(_s, _display, _push);
            AdvancedPage = new PageAdvanced(_s, _display, _push);

            AddPage(WelcomePage, "Decouvrir", "compass");
            AddPage(Display1, "Ecran", "sun");
            AddPage(ColorPage, "Couleur", "palette");
            AddPage(ColorBlindPage, "Daltonisme", "colorblind");
            AddPage(VisionPage, "Vision", "vision");
            AddPage(AutoPage, "Automatisme", "clock");
            AddPage(AppsPage, "Applications", "apps");
            AddPage(ScreensPage, "Ecrans", "screens");
            AddPage(ComfortPage, "Confort", "eye");
            AddPage(HotkeysPage, "Raccourcis", "keyboard");
            AddPage(AdvancedPage, "Avance", "sliders");

            _nav.SelectionChanged += ShowPage;
            LayoutHeader();

            // Le tout premier lancement s'ouvre sur la page qui explique ou
            // l'application se range - sans quoi elle se referme dans la zone de
            // notification et parait avoir disparu. Les lancements suivants ouvrent
            // l'ecran de tous les jours : une explication deja lue qui revient
            // chaque fois devient une porte a pousser.
            ShowPage(_s.IsFirstRun ? 0 : 1);
        }

        private void AddPage(SettingsPage page, string label, string glyph)
        {
            page.Dock = DockStyle.Fill;
            page.Visible = false;
            _content.Controls.Add(page);
            _pages.Add(page);
            _nav.AddTab(label, glyph);
        }

        private void ShowPage(int index)
        {
            if (index < 0 || index >= _pages.Count) return;
            for (int i = 0; i < _pages.Count; i++) _pages[i].Visible = (i == index);
            _pages[index].SyncAndLayout();
            _title.Text = _pages[index].Title;
            _subtitle.Text = _pages[index].Subtitle;
            UpdateStatus();
        }

        // ------------------------------------------------------------------ synchronisation

        /// <summary>Recharge toutes les pages depuis les reglages.</summary>
        public void SyncAll()
        {
            foreach (SettingsPage p in _pages)
            {
                try { p.SyncAndLayout(); } catch { }
            }
            UpdateStatus();
        }

        /// <summary>Recharge seulement la page visible : appelee a chaque changement de valeur.</summary>
        public void RefreshReadouts()
        {
            foreach (SettingsPage p in _pages)
            {
                if (!p.Visible) continue;
                try { p.SyncAndLayout(); } catch { }
            }
            UpdateStatus();
        }

        private void UpdateStatus()
        {
            List<string> bits = new List<string>();

            if (_display.Suspended)
            {
                bits.Add("SUSPENDU - " + _display.SuspendReason);
                _status.ForeColor = Theme.Boost;
            }
            else
            {
                bits.Add(string.Format("{0} % - {1} K",
                    (int)Math.Round(_s.Current.Brightness), _s.Current.Kelvin));
                _status.ForeColor = Theme.Dim;
            }

            // Signale en premier ce qui empeche les reglages de tenir : c'est la cause
            // la plus frequente d'un « ca ne marche pas » attribue a tort a l'application.
            if (IntelDpst.IsAnyDpstActive())
            {
                bits.Add("pilote Intel : luminosite auto active - voir Avance");
                _status.ForeColor = Theme.Boost;
            }
            else if (_display.GammaClamped) bits.Add("Windows rabote les rampes - voir Avance");
            else if (_display.Clipping) bits.Add("hautes lumieres ecretees au-dela de 135 %");
            if (_display.ColorMatrixFailed) bits.Add("matrice plein ecran indisponible");

            _status.Text = string.Join("\n", bits.ToArray());

            _pauseButton.Text = _display.Suspended ? "Reprendre" : "Suspendre";
        }

        // ------------------------------------------------------------------ evenements

        private void OnReset(object sender, EventArgs e)
        {
            _s.Current = new Profile();
            _s.ActiveModeName = "Normal";
            _s.Schedule = ScheduleMode.Manual;
            _s.AdaptiveEnabled = false;
            foreach (MonitorSettings ms in _s.Monitors)
            {
                ms.Blackout = false;
                ms.Enabled = true;
                ms.BrightnessOffset = 0;
            }
            _display.RestoreAll();
            if (_push != null) _push();
            _s.Save();
            SyncAll();
        }

        protected override void OnHandleCreated(EventArgs e)
        {
            base.OnHandleCreated(e);
            DarkTitleBar.Apply(Handle);
        }

        protected override void OnResize(EventArgs e)
        {
            base.OnResize(e);
            LayoutHeader();
        }

        /// <summary>
        /// Le badge d'etat est opaque : il faut donc que titre et sous-titre s'arretent
        /// avant lui, sinon leur fin disparait sous son fond.
        /// </summary>
        private void LayoutHeader()
        {
            if (_header == null || _status == null) return;

            // Les hauteurs viennent des POLICES, et non de nombres ecrits ici :
            // le titre etait fixe a 24 pixels, et le sous-titre commencait a 36.
            // Sur un ecran a 200 %, le titre en demandait 40 : il etait tranche
            // en son milieu et le sous-titre lui passait dessus.
            int statusW = Theme.Px(290);
            int left = Theme.Px(24);
            int hTitre = TextRenderer.MeasureText("Ag", Theme.Title).Height + Theme.Px(2);
            int hSous = TextRenderer.MeasureText("Ag", Theme.Small).Height + Theme.Px(2);
            int hEtat = TextRenderer.MeasureText("Ag", Theme.Small).Height * 2 + Theme.Px(4);

            int haut = Theme.Px(12);
            _header.Height = haut * 2 + hTitre + hSous;

            int statusX = Math.Max(left + Theme.Px(160), _header.Width - statusW - Theme.Px(20));
            _status.SetBounds(statusX, haut, statusW, hEtat);

            int textW = Math.Max(Theme.Px(140), statusX - left - Theme.SpaceLg);
            _title.SetBounds(left, haut, textW, hTitre);
            _subtitle.SetBounds(left, haut + hTitre, textW, hSous);
        }

        /// <summary>Ouvre directement une page donnee (0 = Ecran).</summary>
        public void SelectPage(int index)
        {
            if (_nav != null) _nav.Select(index);
        }

        protected override bool ProcessCmdKey(ref Message msg, Keys keyData)
        {
            // Ctrl+1..9 puis Ctrl+0 : acces direct aux pages, sans quitter le clavier.
            // Le zero sert la dixieme, comme sur la rangee de chiffres du clavier.
            if ((keyData & Keys.Control) == Keys.Control)
            {
                Keys k = keyData & Keys.KeyCode;
                if (k >= Keys.D1 && k <= Keys.D9)
                {
                    _nav.Select(k - Keys.D1);
                    return true;
                }
                if (k == Keys.D0)
                {
                    _nav.Select(9);
                    return true;
                }
            }
            if (keyData == Keys.Escape) { Hide(); return true; }
            return base.ProcessCmdKey(ref msg, keyData);
        }

        protected override void OnFormClosing(FormClosingEventArgs e)
        {
            // La croix range la fenetre ; on ne quitte que par le menu de la zone de notification.
            if (e.CloseReason == CloseReason.UserClosing)
            {
                e.Cancel = true;
                Hide();
            }
            base.OnFormClosing(e);
        }
    }
}

File: mac/Sources/OpusScreen/UI/ControlPanel.swift
import Foundation
import AppKit

/// Fenetre de reglages.
///
/// Onze pages plutot qu'un long formulaire : avec quatre-vingts reglages, tout
/// montrer d'un coup rendrait l'application inutilisable pour le geste quotidien,
/// qui se resume a bouger un curseur. La deuxieme page porte ce geste ; les dix
/// autres attendent qu'on aille les chercher.
///
/// Cette classe n'assemble que la coquille : chaque page connait son domaine et
/// rien d'autre.
public final class ControlPanel: NSObject, NSWindowDelegate {

    public let window: NSWindow
    private let settings: Settings
    private let display: DisplayController
    private let push: () -> Void

    private let root = FlippedView()
    private let nav = SideNav()
    private let content = FlippedView()
    private let titleLabel = LabelView("")
    private let subtitleLabel = LabelView("")
    private let statusLabel = LabelView("")
    private let pauseButton = DarkButton("Suspendre")

    private var pages: [SettingsPage] = []

    public private(set) var welcomePage: PageWelcome!
    public private(set) var displayPage: PageDisplay!
    public private(set) var colorPage: PageColor!
    public private(set) var colorBlindPage: PageColorBlind!
    public private(set) var visionPage: PageVision!
    public private(set) var autoPage: PageAuto!
    public private(set) var appsPage: PageApps!
    public private(set) var screensPage: PageScreens!
    public private(set) var comfortPage: PageComfort!
    public private(set) var hotkeysPage: PageHotkeys!
    public private(set) var advancedPage: PageAdvanced!

    /// Rang de l'onglet Daltonisme, ouvert depuis le menu de la barre des menus.
    ///
    /// Calcule depuis la liste, et non ecrit a la main : une constante avait
    /// ouvert la page voisine le jour ou un onglet s'etait glisse avant elle.
    public var colorBlindPageIndex: Int { pages.firstIndex(where: { $0 === colorBlindPage }) ?? 0 }
    public var visionPageIndex: Int { pages.firstIndex(where: { $0 === visionPage }) ?? 0 }

    /// Declenche la suspension ou la reprise depuis le bouton du pied de page.
    public var togglePause: (() -> Void)?

    private static var navWidth: CGFloat { Theme.px(196) }
    private static var headerHeight: CGFloat { Theme.px(66) }
    private static var footerHeight: CGFloat { Theme.px(52) }

    public init(_ settings: Settings, _ display: DisplayController, _ push: @escaping () -> Void) {
        self.settings = settings
        self.display = display
        self.push = push

        window = NSWindow(contentRect: NSRect(x: 0, y: 0, width: Theme.px(830), height: Theme.px(680)),
                          styleMask: [.titled, .closable, .miniaturizable, .resizable],
                          backing: .buffered, defer: false)
        super.init()

        window.title = "OpusScreen"
        window.appearance = NSAppearance(named: .darkAqua)
        window.backgroundColor = Theme.bg
        window.minSize = NSSize(width: Theme.px(760), height: Theme.px(560))
        window.titlebarAppearsTransparent = true
        window.isReleasedWhenClosed = false
        window.delegate = self
        window.contentView = root

        root.wantsLayer = true
        root.layer?.backgroundColor = Theme.bg.cgColor

        buildUi()
        syncAll()
    }

    // ------------------------------------------------------------------ construction

    private func buildUi() {
        // ---------------- en-tete ----------------
        titleLabel.font = Theme.title
        titleLabel.color = Theme.fg
        root.addSubview(titleLabel)

        subtitleLabel.font = Theme.small
        subtitleLabel.color = Theme.dim
        root.addSubview(subtitleLabel)

        statusLabel.font = Theme.small
        statusLabel.color = Theme.dim
        statusLabel.align = .right
        statusLabel.wraps = true
        root.addSubview(statusLabel)

        // ---------------- navigation ----------------
        // Le nom du produit vit dans la colonne de navigation, pas dans l'en-tete :
        // l'en-tete est ancre a droite du menu et son espace revient au titre de page.
        root.addSubview(nav)

        let brand = LabelView("OpusScreen")
        brand.font = Theme.title
        brand.color = Theme.accent
        brand.frame = NSRect(x: Theme.px(58), y: Theme.px(16),
                             width: ControlPanel.navWidth - Theme.px(64), height: Theme.px(22))
        nav.addSubview(brand)

        let version = LabelView("version " + Installer.currentVersionShort)
        version.font = Theme.small
        version.color = Theme.faint
        version.frame = NSRect(x: Theme.px(59), y: Theme.px(38),
                               width: ControlPanel.navWidth - Theme.px(64), height: Theme.px(16))
        nav.addSubview(version)

        // La place du logo est reservee tout de suite ; l'image arrive quand elle
        // arrive. Attendre une lecture de fichier pour construire une fenetre,
        // c'est accepter que la fenetre n'apparaisse jamais si le systeme decide
        // de poser une question a l'utilisateur d'abord.
        let mark = NSImageView(frame: NSRect(x: Theme.px(14), y: Theme.px(16),
                                             width: Theme.px(38), height: Theme.px(32)))
        mark.imageScaling = .scaleProportionallyUpOrDown
        nav.addSubview(mark)
        AppIcon.withLogo { [weak mark] image in mark?.image = image }

        nav.topOffset = Theme.px(66)
        nav.onSelect = { [weak self] index in self?.showPage(index) }

        // ---------------- pied ----------------
        pauseButton.onClick = { [weak self] in self?.togglePause?() }
        root.addSubview(pauseButton)

        let reset = DarkButton("Tout reinitialiser") { [weak self] in self?.onReset() }
        root.addSubview(reset)
        resetButton = reset

        panicNote.font = Theme.small
        panicNote.color = Theme.faint
        panicNote.align = .right
        panicNote.text = "Secours : ⌃⌥⇧R retablit un ecran normal, meme si l'application ne repond plus."
        root.addSubview(panicNote)

        // ---------------- contenu ----------------
        content.wantsLayer = true
        content.layer?.backgroundColor = Theme.card.cgColor
        root.addSubview(content)

        // ---------------- pages ----------------
        welcomePage = PageWelcome(settings, display, push)
        displayPage = PageDisplay(settings, display, push)
        colorPage = PageColor(settings, display, push)
        colorBlindPage = PageColorBlind(settings, display, push)
        visionPage = PageVision(settings, display, push)
        autoPage = PageAuto(settings, display, push)
        appsPage = PageApps(settings, display, push)
        screensPage = PageScreens(settings, display, push)
        comfortPage = PageComfort(settings, display, push)
        hotkeysPage = PageHotkeys(settings, display, push)
        advancedPage = PageAdvanced(settings, display, push)

        addPage(welcomePage, "Decouvrir", .compass)
        addPage(displayPage, "Ecran", .sun)
        addPage(colorPage, "Couleur", .palette)
        addPage(colorBlindPage, "Daltonisme", .colorblind)
        addPage(visionPage, "Vision", .vision)
        addPage(autoPage, "Automatisme", .clock)
        addPage(appsPage, "Applications", .apps)
        addPage(screensPage, "Ecrans", .screens)
        addPage(comfortPage, "Confort", .eye)
        addPage(hotkeysPage, "Raccourcis", .keyboard)
        addPage(advancedPage, "Avance", .sliders)

        layoutChrome()

        // Le tout premier lancement s'ouvre sur la page qui explique ou
        // l'application se range - sans quoi elle se referme dans la barre des
        // menus et parait avoir disparu. Les lancements suivants ouvrent l'ecran
        // de tous les jours : une explication deja lue qui revient chaque fois
        // devient une porte a pousser.
        let first = settings.isFirstRun ? 0 : 1
        nav.setSelectionSilent(first)
        showPage(first)
    }

    private var resetButton: DarkButton?
    private let panicNote = LabelView("")

    private func addPage(_ page: SettingsPage, _ label: String, _ glyph: NavGlyph) {
        page.isHidden = true
        content.addSubview(page)
        pages.append(page)
        nav.addTab(label, glyph)
    }

    private func showPage(_ index: Int) {
        guard index >= 0, index < pages.count else { return }
        for (i, p) in pages.enumerated() { p.isHidden = (i != index) }
        pages[index].frame = content.bounds
        pages[index].syncAndLayout()
        titleLabel.text = pages[index].pageTitle
        subtitleLabel.text = pages[index].pageSubtitle
        updateStatus()
    }

    public func selectPage(_ index: Int) {
        nav.select(index)
    }

    // ------------------------------------------------------------------ mise en page

    private func layoutChrome() {
        let w = root.bounds.width
        let h = root.bounds.height
        guard w > 0, h > 0 else { return }

        let header = ControlPanel.headerHeight
        let footer = ControlPanel.footerHeight
        let navW = ControlPanel.navWidth

        nav.frame = NSRect(x: 0, y: 0, width: navW, height: h)

        let statusWidth = Theme.px(300)
        let left = navW + Theme.px(24)
        let statusX = max(left + Theme.px(160), w - statusWidth - Theme.px(20))

        titleLabel.frame = NSRect(x: left, y: Theme.px(14),
                                  width: max(Theme.px(140), statusX - left - Theme.spaceLg),
                                  height: Theme.px(22))
        subtitleLabel.frame = NSRect(x: left, y: Theme.px(36),
                                     width: max(Theme.px(140), statusX - left - Theme.spaceLg),
                                     height: Theme.px(18))
        statusLabel.frame = NSRect(x: statusX, y: Theme.px(14), width: statusWidth, height: Theme.px(40))

        content.frame = NSRect(x: navW, y: header, width: w - navW, height: h - header - footer)
        for p in pages where !p.isHidden {
            p.frame = content.bounds
            p.relayout()
        }

        pauseButton.frame = NSRect(x: navW + Theme.px(20), y: h - footer + Theme.px(8),
                                   width: Theme.px(116), height: Theme.minTarget)
        resetButton?.frame = NSRect(x: navW + Theme.px(144), y: h - footer + Theme.px(8),
                                    width: Theme.px(140), height: Theme.minTarget)
        panicNote.frame = NSRect(x: navW + Theme.px(292), y: h - footer + Theme.px(8),
                                 width: max(Theme.px(80), w - navW - Theme.px(312)),
                                 height: Theme.minTarget)
    }

    public func windowDidResize(_ notification: Notification) {
        layoutChrome()
    }

    // ------------------------------------------------------------------ synchronisation

    /// Recharge toutes les pages depuis les reglages.
    public func syncAll() {
        for p in pages { p.syncAndLayout() }
        updateStatus()
    }

    /// Recharge seulement la page visible : appelee a chaque changement de valeur.
    public func refreshReadouts() {
        for p in pages where !p.isHidden { p.syncAndLayout() }
        updateStatus()
    }

    private func updateStatus() {
        var bits: [String] = []

        if display.suspended {
            bits.append("SUSPENDU - " + display.suspendReason)
            statusLabel.color = Theme.boost
        } else {
            bits.append("\(Int(settings.current.brightness.rounded())) % - \(settings.current.kelvin) K")
            statusLabel.color = Theme.dim
        }

        // Signale en premier ce qui empeche les reglages de tenir : c'est la cause
        // la plus frequente d'un « ca ne marche pas » attribue a tort a l'application.
        if AutoBrightness.isAnyAutoActive() {
            bits.append("luminosite automatique active - voir Avance")
            statusLabel.color = Theme.boost
        } else if display.gammaClamped {
            bits.append("table de couleurs rabotee - voir Avance")
        } else if display.clipping {
            bits.append("hautes lumieres ecretees au-dela de 135 %")
        }
        if display.colorMatrixFailed { bits.append("matrice de couleur indisponible") }

        statusLabel.text = bits.joined(separator: "\n")
        pauseButton.title = display.suspended ? "Reprendre" : "Suspendre"
    }

    // ------------------------------------------------------------------ evenements

    private func onReset() {
        settings.current = Profile()
        settings.activeModeName = "Normal"
        settings.schedule = .manual
        settings.adaptiveEnabled = false
        for ms in settings.monitors {
            ms.blackout = false
            ms.enabled = true
            ms.brightnessOffset = 0
        }
        display.restoreAll()
        push()
        settings.save()
        syncAll()
    }

    public func show() {
        NSApp.activate(ignoringOtherApps: true)
        window.makeKeyAndOrderFront(nil)
        layoutChrome()
        syncAll()
    }

    /// La croix range la fenetre ; on ne quitte que par le menu de la barre des menus.
    public func windowShouldClose(_ sender: NSWindow) -> Bool {
        window.orderOut(nil)
        return false
    }

    /// ⌘1 a ⌘9 puis ⌘0 : acces direct aux pages, sans quitter le clavier.
    /// Le zero sert la dixieme, comme sur la rangee de chiffres du clavier.
    public func handleKey(_ event: NSEvent) -> Bool {
        if event.modifierFlags.contains(.command), let chars = event.charactersIgnoringModifiers {
            if let digit = Int(chars), digit >= 1, digit <= 9 {
                nav.select(digit - 1)
                return true
            }
            if chars == "0" { nav.select(9); return true }
        }
        if event.keyCode == 53 {   // echappement
            window.orderOut(nil)
            return true
        }
        return false
    }
}
